### ISSUE TITLE
Cannot set labels with non-ASCII characters in exFAT

### ISSUE THREAD

**Issue by bertogg** [storaged-project/libblockdev#846]
I reproduced this in UDisks but I think that the problem is here. I'm using the master branch in both cases. I confirm that the problem does not happen with the 2.x stable releases.

exFAT filesystems are formatted using `mkfs.exfat` from the `exfatprogs` package. This volume label in this filesystem is a Unicode string ([exFAT spec, section 7.3.3](https://learn.microsoft.com/en-gb/windows/win32/fileio/exfat-specification#733-volumelabel-field)) encoded in UTF-16, but `mkfs.exfat` uses `mbstowcs()` to do the conversion ([here](https://github.com/exfatprogs/exfatprogs/blob/1.2.0/lib/libexfat.c#L234)), which depends on `LC_CTYPE` from the current locale to determine the encoding of the string provided by the user.

However, and even if the OS supports UTF-8 correctly, when libblockdev spawns a process it sets `LC_ALL` to `C` ([here](https://github.com/storaged-project/libblockdev/blob/4a85c9d44f0cd402c1bf5c9ae53d646a3a358fa8/src/utils/exec.c#L215) and [here](https://github.com/storaged-project/libblockdev/blob/4a85c9d44f0cd402c1bf5c9ae53d646a3a358fa8/src/utils/exec.c#L419)), making the conversion fail if the label contains non-ASCII characters ("invalid character sequence in current locale").

**Comment by bertogg** [comment]
If this is fixed then [this test case](https://github.com/storaged-project/udisks/blob/9af2b726d637572936255e197f27465ac3eb22d5/src/tests/dbus-tests/test_80_filesystem.py#L952) can be updated.

**Comment by tbzatek** [comment]
> However, and even if the OS supports UTF-8 correctly, when libblockdev spawns a process it sets `LC_ALL` to `C` ([here](https://github.com/storaged-project/libblockdev/blob/4a85c9d44f0cd402c1bf5c9ae53d646a3a358fa8/src/utils/exec.c#L215) and [here](https://github.com/storaged-project/libblockdev/blob/4a85c9d44f0cd402c1bf5c9ae53d646a3a358fa8/src/utils/exec.c#L419)), making the conversion fail if the label contains non-ASCII characters ("invalid character sequence in current locale").

Agree this could be a problem. Shall we break the libblockdev exec API for 3.0 @vojtechtrefny?

UDisks is heavily based on UTF-8 and so does the D-Bus protocol for `s` data type which is used for any label in UDisks.

**Comment by bertogg** [comment]
Would setting `LC_ALL` to `C.UTF-8` be an option? I don't think that needs any API changes.... `C.UTF-8` is available since glibc 2.35 although it doesn't necessary mean that it's enabled (in practice it's probably unlikely that it isn't...)

**Comment by bertogg** [comment]
> What happens when C.UTF-8 is not available? Would the very fallback of C locale be used instead? A quick test in bash indicates so, so this should guarantee some compatibility.

I'm not completely sure, I can test it a bit more carefully.

If that doesn't work we can also try to detect if `C.UTF-8` is available and use `C` otherwise.

Option B (maybe the best one) is to make the default locale configurable at build time, like e.g. systemd did recently: https://github.com/systemd/systemd/commit/c9bd01ae902d66acd8647a1cc0167bc0d246cbc8 , so default to the UTF-8 one (which is a sensible choice for modern distros) but let downstreams change it if it doesn't work for them.

**Comment by bertogg** [comment]
Oh and I forgot, both CentOS 8 and 9 (at least) patch glibc to add the C.UTF-8 locale:

https://gitlab.com/redhat/centos-stream/rpms/glibc/-/blob/c8s/glibc-c-utf8-locale.patch
https://gitlab.com/redhat/centos-stream/rpms/glibc/-/blob/c9s/glibc-c-utf8-locale-2.patch

And so does Debian and derivatives (this thing was invented there).

But I'm not sure that it's enabled by default in Arch and derivatives, even with recent glibc versions.

**Comment by bertogg** [comment]
If we use C.UTF-8, is this bug relevant to libblockdev? https://sourceware.org/bugzilla/show_bug.cgi?id=16621

**Comment by vojtechtrefny** [comment]
> If we use C.UTF-8, is this bug relevant to libblockdev? https://sourceware.org/bugzilla/show_bug.cgi?id=16621

Yes, that's definitely a problem. We use `LC_ALL=C` because we often parse command output, not with `mkfs`, so we could use `C.UTF-8` when running `mkfs.exfat`, but we can't switch it globally.

**Comment by tbzatek** [comment]
So what if we redefine all the locale variables?
```
$ LC_ALL="C.UTF-8" LANG=$LC_ALL locale
LANG=C.UTF-8
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=C.UTF-8
```

**Comment by bertogg** [comment]
Since what matters for `mbstowcs()` is the value of `LC_CTYPE` (_"This specifies what locale to use for character sets and character classification"_) I think what's important is that is the variable set to an UTF-8 locale. Then `LC_MESSAGES` can be set to `C` if necessary.

Or we simply unset `LANGUAGE`, which is what takes precedence over `LC_ALL=C.UTF-8`

```
$ LANGUAGE=fr_FR LC_ALL=C.UTF-8 cp
cp: opérande de fichier manquant

$ LANGUAGE=fr_FR LC_CTYPE=C.UTF-8 LC_MESSAGES=C cp
cp: missing file operand

$ LANGUAGE="" LC_ALL=C.UTF-8 cp
cp: missing file operand
```

**Comment by bertogg** [comment]
> So what if we redefine all the locale variables?

From [the glibc manual](https://www.gnu.org/software/libc/manual/2.37/html_node/Locale-Categories.html): _`LC_ALL` [...] _overwrites all selections by the other `LC_*` variables or `LANG`._

I think it's enough by setting `LC_ALL` and unsetting `LANGUAGE`, which is what takes precedence over all the others, from [the manual](https://www.gnu.org/software/libc/manual/2.37/html_node/Using-gettextized-software.html), the order is:

```
LANGUAGE
LC_ALL
LC_xxx
LANG
```

**Comment by bertogg** [comment]
> What happens when C.UTF-8 is not available?

I've been testing this a bit more carefully today.

When a child program is spawned it starts with the locale set to `C` so the child first needs to set the locale, typically with `setlocale(LC_xxx, "")`, where the empty string means _"get the name of the new locale from the environment"_.

This means: check `LC_ALL`, then `LC_xxx`, then `LANG`, otherwise use `C`. If the value is set but is invalid (`C.UTF-8` if it doesn't exist) then the call fails and the locale is unmodified. _This happens even if the variables with lower priority than the incorrect one have a valid locale._

So if `LC_ALL=C.UTF-8` and `setlocale()` fails because `C.UTF-8` does not exist in the system, then `C` is guaranteed unless the locale was changed in an ealier call. But since we're talking about a new child process querying the environment on startup to initialize the locale then this won't happen.

Finally, and as I explained in my previous comment, gettext will give priority to the `LANGUAGE` variable if the locale is not `C`, so we have to unset it.

Because of this, I think that the changes are safe. I just created a pull request: https://github.com/storaged-project/libblockdev/pull/847
